flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java

```diff
@@ -0,0 +1,197 @@
+                        StringUtils.join(unfinishedExecutionStates, ", "), jobId);
+        FlinkException inconsistentFinalStateException = new FlinkException(errorMessage);
+
+        scheduler.handleGlobalFailure(inconsistentFinalStateException);
+        return terminateExceptionally(inconsistentFinalStateException);
+    }
+
+    private StopWithSavepointState terminateExceptionally(Throwable throwable) {
+        scheduler.startCheckpointScheduler(checkpointCoordinator);
+        result.completeExceptionally(throwable);
+
+        return StopWithSavepointState.Final;
+    }
+
+    private StopWithSavepointState terminateSuccessfully(String path) {
+        result.complete(path);
+
+        return StopWithSavepointState.Final;
+    }
+
+    private static Set<ExecutionState> extractUnfinishedStates(
+            Collection<ExecutionState> executionStates) {
+        return executionStates.stream()
+                .filter(state -> state != ExecutionState.FINISHED)
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * {@code StopWithSavepointState} represents the different states during the stop-with-savepoint
+     * operation.
+     *
+     * <p>The state transitions are implemented in the following way: InitialWait ->
+     * [WaitForSavepointCreation|WaitForJobTermination] -> Final
+     */
+    private enum StopWithSavepointState {
+        InitialWait {
+            @Override
+            protected StopWithSavepointState handleSavepointCreation(
+                    StopWithSavepointContext context, String path) {
+                context.path = path;
+                return WaitForJobTermination;
+            }
+
+            @Override
+            public StopWithSavepointState onExecutionsTermination(
+                    StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
+                context.unfinishedStates = extractUnfinishedStates(executionStates);
+                return WaitForSavepointCreation;
+            }
+        },
+        WaitForSavepointCreation {
+            @Override
+            public StopWithSavepointState handleSavepointCreation(
+                    StopWithSavepointContext context, String path) {
+                Preconditions.checkState(
+                        context.unfinishedStates != null,
+                        InitialWait + " should have preceded: No unfinishedStates is set.");
+
+                if (!context.unfinishedStates.isEmpty()) {
+                    return context.terminateExceptionWithGlobalFailover(context.unfinishedStates);
+                }
+
+                return context.terminateSuccessfully(path);
+            }
+        },
+        WaitForJobTermination {
+            @Override
+            public StopWithSavepointState onExecutionsTermination(
+                    StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
```

Comment:
I am wondering whether the `context` really helps in the overall design. It seems that we use the context for storing state and to trigger some methods. Making the state an inner class could solve the latter problem. Using classes instead of enums could then allow us to store the state with the state instance instead of some fields in the outer class which are never used in the outer context.

Reply:
I refactored it to use classes now. 👍